File: 0207-course-schedule/0207-course-schedule.cpp
class Solution {
public:
    
    bool dfs(vector<vector<int>>&adj,int v,vector<int>&vis)
    {
        vis[v] = 1;
        for(int u:adj[v])
        {
            if(vis[u] == 0)
            {
                if(!dfs(adj,u,vis))
                {
                    return false;
                }
            }
            else if(vis[u] == 1)
            {
                return false;
            }
        }
        vis[v] = 2;
        return true;
        
    }
    bool canFinish(int numCourses, vector<vector<int>>& pre) {
        vector<vector<int>>adj(numCourses);
        vector<int>vis(numCourses,0);
        for(auto p:pre)
        {
            adj[p[1]].push_back(p[0]);
        }
        for(int i=0;i<numCourses;i++)
        {
            if(vis[i] == 0)
            {
                if(!dfs(adj,i,vis))
                {
                    return false;
                }
            }
        }
        //cout << res.size() << "\n";
        return true;
    }
};

File: 0207-course-schedule/0207-course-schedule.py
from collections import defaultdict,deque
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        adj = defaultdict(list)
        indeg = [0]*numCourses
        for u,v in prerequisites:
            adj[v].append(u)
            indeg[u] += 1
        q = deque()
        for i in range(numCourses):
            if indeg[i] == 0:
                q.append(i)
        took = 0
        while q:
            took += 1
            fr = q.popleft()
            for neigh in adj[fr]:
                indeg[neigh] -= 1
                if indeg[neigh] == 0:
                    q.append(neigh)
        return numCourses == took
        

        
        